### ISSUE TITLE
[Feature] Expose a formal programatic API

### ISSUE THREAD

**Issue by jpnelson** [lerna/lerna#167]
@thejameskyle asked me to compile this list of things we'd need to expose for a public API, for https://github.com/atlassian/lerna-semantic-release
- We need a way to know what all of the packages are (currently using `PackageUtilities.getPackages`). We need to be able to get the `package.json` data from the packages.

Other than that, lerna is mostly used through the CLI – you `lerna run` most of the semantic release steps. The only other things necessary have been added in PR's, such as https://github.com/lerna/lerna/pull/163

**Comment by gigabo** [comment]
lerna-changelog is currently using:

- "lerna/lib/Repository" for `lernaJson` (config object) and `rootPath` (repo root)
- "lerna/lib/progressBar" just to present a lerna-like UI feel

Wish list:

- Would _like_ to get at a `Package` object, ideally given just a package name.  Doesn't look like such an interface exists even internally currently.  Something simple like: `lerna.getPackageByName(<name>)`.  Want to use this to [check `package.isPrivate()`](https://github.com/lerna/lerna-changelog/issues/15).

**Comment by seansfkelley** [comment]
#174 might be a good forcing function for this hypothetical API.

**Comment by ahfarmer** [comment]
I currently use the non-formal programmatic API to find out which packages have changed on CI, and only test those packages (for speed reasons).

Would love it if it was a formal API. What I'm doing is not ideal:

```
const logger = require('lerna/lib/logger');
const UpdatedPackagesCollector = require('lerna/lib/UpdatedPackagesCollector');
const Repository = require('lerna/lib/Repository');
logger.setLogLevel('silent');

const repo = new Repository();
const collector = new UpdatedPackagesCollector(repo, {}, {});
console.log(collector.collectUpdatedPackages()); // but I don't actually log them - I use this information
```

**Comment by lukebatchelor** [comment]
@ahfarmer That is literally exactly what I'm about to do.

@gigabo FYI, I'm looking at setting this up in our CI and then pushing changes back to allow this. Is this still something you think we'd like in lerna? Do you have a preferred way of exposing functions?

**Comment by Alxandr** [comment]
Any news on this? Will it be available for 2.0?

**Comment by evocateur** [comment]
@Alxandr It's not really gonna happen until #174 picks up some steam. The subpath imports _technically_ work now, but I wouldn't recommend that as subpath imports are essentially abusing what amounts to a private API, and is thus not guaranteed by semver (that is, it could break at any patch or minor version).
